source/system/structs/HashMap.ooc

```diff
@@ -6,85 +6,50 @@
  * of the MIT license.  See the LICENSE file for details.
  */
 
-HashEntry: cover {
-	key: Pointer
-	value: Pointer
-	next: This* = null
-	init: func@ { this init(null, null) }
-	init: func@ ~keyVal (=key, =value)
-	free: func {
-		if (this key != null)
-			memfree(this key)
-		if (this value != null)
-			memfree(this value)
-		if (this next != null) {
-			temp := this next@
-			temp free()
-			memfree(this next)
+_HashEntry: class <K, V> {
+	_key: __onheap__ K
+	_value: __onheap__ V
+	_next: This<K, V> = null
+	init: func (=_key, =_value)
+	free: override func {
+		memfree(this _key)
+		memfree(this _value)
+		if (this _next != null) {
+			this _next free()
+			this _next = null
 		}
+		super()
 	}
 }
 
-intHash: func <K> (key: K) -> SizeT { (key as Int) as SizeT }
-pointerHash: func <K> (key: K) -> SizeT { (key as Pointer) as SizeT }
-
-/**
-	Port of Austin Appleby's Murmur Hash implementation
-	http://code.google.com/p/smhasher/
-*/
-murmurHash: func <K> (keyTagazok: K) -> SizeT {
-	seed: SizeT = 1
-
-	len := K size
-	m = 0x5bd1e995: SizeT
-	r = 24: SSizeT
-	l := len
-
-	h : SizeT = seed ^ len
-	data := (keyTagazok&) as Byte*
-
+// Austin Appleby's Murmur Hash implementation: http://code.google.com/p/smhasher/
+_murmurHash: func <K> (key: K) -> Int {
+	(len, m, r, l, h, t) := (K size, 0x5bd1e995 as SizeT, 24 as SizeT, len, (1 as SizeT) ^ len as SizeT, 0)
+	data := (key&) as Byte*
 	while (true) {
 		k := (data as SizeT*)@
-
 		k *= m
 		k ^= k >> r
 		k *= m
-
 		h *= m
 		h ^= k
-
 		data += 4
 		if (len < 4) break
 		len -= 4
 	}
-
-	t := 0
-
 	if (len == 3) h ^= data[2] << 16
 	if (len == 2) h ^= data[1] << 8
 	if (len == 1) h ^= data[0]
-
 	t *= m; t ^= t >> r; t *= m; h *= m; h ^= t
 	l *= m; l ^= l >> r; l *= m; h *= m; h ^= l
-
-	h ^= h >> 13
-	h *= m
-	h ^= h >> 15
-
-	h
+	h ^= h >> 13; h *= m; h ^= h >> 15
+	h as Int
 }
 
-/**
- * khash's ac_X31_hash_string
- * http://attractivechaos.awardspace.com/khash.h.html
- * @access private
- * @param s The string to hash
- * @return UInt
- */
-acX31Hash: func <K> (key: K) -> SizeT {
-	raise(key == null, "key in acX31Hash must be non-null!")
-	s : Char* = (K == String) ? (key as String) toCString() as Char* : key as Char*
-	h = s@ : SizeT
+// khash's ac_X31_hash_string: http://attractivechaos.awardspace.com/khash.h.html
+_acX31Hash: func <K> (key: K) -> Int {
```

Comment:
I don't know enough about hashing functions to decide, but maybe this function is enough for general purposes instead of murmur...